feat(ast): allow conversion from `Expression` into `Statement` with `FromIn` trait.

rzvxa opened the pull request:
Our downstream rolldown is using this, If we want them to adopt our traits we should provide this implementation for them.

https://github.com/rolldown/rolldown/blob/2cd2a367ee25c3d1f711fda9b1d3ead5982cfcff/crates/rolldown_ecmascript/src/allocator_helpers/into_in.rs#L24

rzvxa commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4124?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4124** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4124?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#4122** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4122?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#4117** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4117?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#4116** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4116?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#3890** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3890?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#4098** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4098?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#4090** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4090?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#4088** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4088?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rzvxa and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_ast/src/ast_impl/js.rs:
```diff
@@ -665,6 +665,15 @@ impl<'a> Statement<'a> {
 
+impl<'a> FromIn<'a, Expression<'a>> for Statement<'a> {
+    fn from_in(expression: Expression<'a>, alloc: &'a oxc_allocator::Allocator) -> Self {
+        Statement::ExpressionStatement(Box::from_in(
```
I would suggest using `Box::new_in` here. It's equivalent, but a little more clear.

rzvxa replied:
I explicitly used `from_in` here to keep the propagation within the `FromIn` trait implementers. This way if any major change happens to the trait it will propagate through the implementations. That's exactly why I prefer to use `IntoIn` if I were implementing an `IntoIn`. So let's say for example for some unknown reason, we would want to accept an extra parameter or maybe change the `Allocator` to `AllocationCtx`; We can migrate more easily with this pattern.

I'm open to changing it since at this point it makes no difference whatsoever. I just wanted to let you know my thought process while I was coming up with this specific implementation.

rzvxa committed:
feat(ast): allow conversion from `Expression` into `Statement` with `FromIn` trait. (#4124)

Our downstream rolldown is using this, If we want them to adopt our traits we should provide this implementation for them.

https://github.com/rolldown/rolldown/blob/2cd2a367ee25c3d1f711fda9b1d3ead5982cfcff/crates/rolldown_ecmascript/src/allocator_helpers/into_in.rs#L24
